sky/services/ns_net/url_loader_impl.mm

```diff
@@ -95,6 +98,57 @@ - (void)dealloc {
 
 namespace mojo {
 
+class AsyncNSDataDrainer : common::DataPipeDrainer::Client {
+ public:
+  using CompletionCallback = base::Callback<void(NSData* /* transfer-none */)>;
+
+  AsyncNSDataDrainer()
+      : data_([[NSMutableData alloc] init]), draining_(false) {}
+
+  void StartWithCompletionCallback(ScopedDataPipeConsumerHandle source,
+                                   CompletionCallback callback) {
+    DCHECK(!draining_)
+        << "A instance of an AsyncNSDataDrainer can only be used to drain once";
+    DCHECK(drainer_ == nullptr);
+
+    draining_ = true;
+    callback_ = callback;
+
+    // There is no "Start" method on a data pipe drainer. Instantiating
+    // an instance automatically starts the drain.
+    drainer_ = std::unique_ptr<common::DataPipeDrainer>(
+        new common::DataPipeDrainer(this, source.Pass()));
+  }
+
+  ~AsyncNSDataDrainer() override { [data_ release]; }
+
+ private:
+  CompletionCallback callback_;
+  NSMutableData* data_;
+  std::unique_ptr<common::DataPipeDrainer> drainer_;
+  bool draining_;
+
+  void OnDataAvailable(const void* data, size_t num_bytes) override {
+    [data_ appendBytes:data length:num_bytes];
+  }
+
+  void OnDataComplete() override {
+    auto callback = callback_;
+    auto data = data_;
```

Comment:
I would put the type in here explicitly because it's important that its an `NSMutableData* ` rather than a `NSMutableData*&`, for example.